### PR TITLE
Automation for Re design of error page.

### PR DESCRIPTION
**Pull request by labkey-sweta**
#### Rationale
LabKey R&D UX - Re-design for error/page not found pages - Automation coverage.

#### Related Pull Requests
None

#### Changes
Two new files LabkeyErrorPageTest.java and LabkeyErrorPage.java

### CODE REVIEW

File: src/org/labkey/test/tests/LabkeyErrorPageTest.java [hunk 1]
```diff
@@ -0,0 +1,135 @@
...
+        LabkeyErrorPage errorPage = new LabkeyErrorPage(getDriver());
+
+        checker().verifyEquals("Incorrect error heading message","Oops! The requested page cannot be found.",
+                errorPage.getErrorHeading());
+        checker().verifyEquals("Incorrect error sub-heading message","This is a test message for not found exception.",
+                errorPage.getSubErrorHeading());
+        checker().verifyTrue("Incorrect error image",errorPage.getErrorImage().contains(imageTitle));
+
+        goToProjectHome();
+        beginAt(WebTestHelper.buildRelativeUrl("project",getCurrentContainerPath(),"beginning"));
```
Comment (labkey-tchad) [inline src/org/labkey/test/tests/LabkeyErrorPageTest.java]: Remove unnecessary navigation
```suggestion
        beginAt(WebTestHelper.buildRelativeUrl("project",getCurrentContainerPath(),"beginning"));
```

**Review by labkey-tchad — CHANGES_REQUESTED** [review]
We should get rid of the `Throw*Action`s from `MothershipController` and use the ones in `TestController`. There are existing test helpers for them (`org.labkey.test.pages.test.TestActions.ExceptionActions`) and they allow additional customization. Specifically, some allow custom messages which would be good for checking for script injection.

File: src/org/labkey/test/tests/LabkeyErrorPageTest.java [hunk 2]
```diff
@@ -0,0 +1,135 @@
...
+    }
+
+    private void doSetup()
+    {
+        _containerHelper.createProject(getProjectName(), null);
+        createUserWithPermissions(READER_USER, getProjectName(), "Reader");
+    }
+
+    @Test
+    public void testGeneralErrors()
```
Comment (labkey-ankurj) [inline src/org/labkey/test/tests/LabkeyErrorPageTest.java:37]: I'd suggest to add one more test case for the incorrect controller name.

**Review by labkey-danield — APPROVED** [review]
I have nothing to add on top of Ankurs and Treys comments. 

I like how you have organized the tests and made each one narrow in scope!

File: src/org/labkey/test/tests/LabkeyErrorPageTest.java (hunk 1 above)
Comment (labkey-sweta) [reply src/org/labkey/test/tests/LabkeyErrorPageTest.java]: Removed

File: src/org/labkey/test/tests/LabkeyErrorPageTest.java (hunk 2 above)
Comment (labkey-sweta) [reply src/org/labkey/test/tests/LabkeyErrorPageTest.java:37]: done

**Review by labkey-ankurj — APPROVED** [review]

File: src/org/labkey/test/tests/LabkeyErrorPageTest.java [hunk 3]
```diff
@@ -0,0 +1,134 @@
...
+        String imageTitle = "notFound_error.svg";
+        log("Verifying error message with mothership controller and action");
+        beginAt(WebTestHelper.buildURL("test", "NotFound"));
+        LabkeyErrorPage errorPage = new LabkeyErrorPage(getDriver());
+
+        checker().verifyEquals("Incorrect error heading message","Oops! The requested page cannot be found.",
+                errorPage.getErrorHeading());
+        checker().verifyEquals("Incorrect error sub-heading message","404: page not found.",
+                errorPage.getSubErrorHeading());
+        checker().verifyTrue("Incorrect error image",errorPage.getErrorImage().contains(imageTitle));
```
Comment (labkey-tchad) [inline src/org/labkey/test/tests/LabkeyErrorPageTest.java]: Use hamcrest assertions to provide more informative error messages.
```suggestion
        checker().verifyThat("Incorrect error image", errorPage.getErrorImage(), CoreMatchers.containsString(imageTitle));
```

**Review by labkey-tchad — APPROVED** [review]
I suggest replacing most of the `verifyTrue` calls with `verifyThat` calls to provide more informative error messages if they fail.

File: src/org/labkey/test/tests/LabkeyErrorPageTest.java (hunk 3 above)
Comment (labkey-sweta) [reply src/org/labkey/test/tests/LabkeyErrorPageTest.java]: done

**Pull request merged**